### PR TITLE
Fix `diff()` to support OffsetArrays

### PR DESCRIPTION
**Pull request by ghost**
Fix JuliaArrays/OffsetArrays.jl#79

### CODE REVIEW

**Review by timholy — COMMENTED** [review]
This is very good, many thanks for doing it! There's just one key question we should answer: should the axes of the output be `Base.OneTo` or something else? (This implementation uses `Base.OneTo`.) The only other consistent option I see is to use `typeof(ax)(first(ax):last(ax)-1)`, so that for this example the axes of the result would be
```julia
Base.IdentityUnitRange.(-1:0, -1:1)
```
and you'd compare against `OffsetArray(fill(1, 2, 4), -2, -2)`.

"Narrowing" from the end is consistent with the existing behavior for `::Array`, which is why I went with that option.

**Comment by mikan2087** [comment]
@timholy Thank you for reviewing and showing another option. I think the "Narrowing" option is better, so I fixed the code and the test case.

File: base/multidimensional.jl [hunk 1]
```diff
@@ -823,14 +823,13 @@ julia> diff(vec(a))
...
     1 <= dims <= N || throw(ArgumentError("dimension $dims out of range (1:$N)"))
 
     r = axes(a)
-    r0 = ntuple(i -> i == dims ? UnitRange(1, last(r[i]) - 1) : UnitRange(r[i]), N)
-    r1 = ntuple(i -> i == dims ? UnitRange(2, last(r[i])) : UnitRange(r[i]), N)
+    r0 = ntuple(i -> i == dims ? UnitRange(first(r[i]), last(r[i]) - 1) : UnitRange(r[i]), N)
+    r1 = ntuple(i -> i == dims ? UnitRange(first(r[i]) + 1, last(r[i])) : UnitRange(r[i]), N)
 
-    return view(a, r1...) .- view(a, r0...)
+    return reshape(view(a, r1...) .- view(a, r0...), r0)
```
Comment (timholy) [inline base/multidimensional.jl]: This `reshape` is causing test failures, e.g., `MethodError: no method matching reshape(::Array{Float64,2}, ::Tuple{UnitRange{Int64},UnitRange{Int64}})` (that method is defined only if you've loaded `OffsetArrays`).

In principle, you don't need `reshape` if you pay attention to the indices-of-the-indices. That is, `r0` and `r1` themselves have `Base.OneTo` axes, which is why the difference `view1 .- view2` has `Base.OneTo` axes and thus needs a `reshape`. But instead, if you make sure that they preserve their own axes, then the `view1 .- view2` will use those axes.

For example:
```julia
trimlast(ax::OneTo) = OneTo(ax.stop-1)
trimlast(ax::UnitRange) = UnitRange(ax.start, ax.stop-1)
trimlast(ax::IdentityUnitRange) = IdentityUnitRange(trimlast(ax.indices))

r0 = ntuple(i -> i == dims ? trimlast(r[i]) : r[i], N)
r1 = ntuple(i -> i == dims ? trimlast(r[i]) .+ 1 : r[i], N)

view(a, r1...) .- view(a, r0...)
```
seems like it should work, except that OffsetArrays will need to define that broadcasting operation for `IdentityUnitRange`.

**Review by timholy — CHANGES_REQUESTED** [review]

**Comment by DilumAluthge** [comment]
It looks like this pull request is no longer mergeable.

If you would still like to make this change, please make a new pull request based on the latest `master` branch of Julia.

**Comment by LilithHafner** [comment]
Another option for result axes is fractional:

If the input axes are `10:20` then the output axes are `10.5:19.5`. This would be sensible for computer vision but inconsistent with current behavior when axes are one based.

**Pull request closed without merging**